cloud/shared/bin/lib/setup.py

```diff
@@ -34,22 +34,57 @@
 
 template_dir = config_loader.get_template_dir()
 Setup = load_class(template_dir)
-template_setup = Setup()
+template_setup = Setup(config_loader)
+template_setup.pre_terraform_setup()
+backend_vars = "backend_vars"
+setup_terraform_variables = template_setup.get_setup_terraform_variables()
 
 ###############################################################################
 # Terraform Init/Plan/Apply
 ###############################################################################
 
-terraform_tfvars_filename = f"{template_dir}/setup.auto.tfvars"
+terraform_tfvars_filename = "setup.auto.tfvars"
+terraform_tfvars_path = f"{template_dir}/{terraform_tfvars_filename}"
 
 # Write the passthrough vars to a temporary file
-tf_var_writter = TfVarWriter(terraform_tfvars_filename)
-variables_to_write = config_loader.get_terraform_variables()
-tf_var_writter.write_variables(variables_to_write)
-
-# TODO need to call input/apply like how the setup script in staging_azure
-subprocess.call(
-    f"terraform apply -input=false -var-file={terraform_tfvars_filename}",
-    shell=True,
-    cwd=template_dir,
-)
+tf_var_writter = TfVarWriter(terraform_tfvars_path)
+conf_variables = config_loader.get_terraform_variables()
+tf_var_writter.write_variables({
+    **conf_variables,
+    **setup_terraform_variables
+})
+
+# Note that the -chdir means we use the relative paths for 
+# both the backend config and the var file
+subprocess.check_call([
+    "terraform", 
+    f"-chdir={template_dir}", 
+    "init", 
+    f"-backend-config={backend_vars}"
+])
+
+subprocess.check_call([
+    "terraform", 
+    f"-chdir={template_dir}", 
+    "apply", 
+    f"-var-file={terraform_tfvars_filename}"
+])
+
+###############################################################################
+# Post Run Setup Tasks (if needed)
+###############################################################################
+
+if template_setup.requires_post_terraform_setup():
+    template_setup.post_terraform_setup()
+    setup_terraform_variables = template_setup.get_setup_terraform_variables()
+    tf_var_writter.write_variables({
+        **conf_variables, 
+        **setup_terraform_variables
+    })
```

Comment:
instead of generating these and putting them in the tf vars file, we should generate and push them into key vault

Reply:
will move the adfs client into secrets in the follow up pr to try to keep this somehwat smaller